x/superfluid/keeper/stake.go

```diff
@@ -249,47 +242,75 @@ func (k Keeper) SuperfluidDelegate(ctx sdk.Context, sender string, lockID uint64
 	return k.mintOsmoTokensAndDelegate(ctx, amount, acc)
 }
 
-// SuperfluidUndelegate starts undelegating superfluid delegated position for the given lock.
-// Undelegation is done instantly and the equivalent amount is sent to the module account
-// where it is burnt instantly. Note that this method does not include unbonding the lock
-// itself.
-func (k Keeper) SuperfluidUndelegate(ctx sdk.Context, sender string, lockID uint64) error {
+// undelegateCommon is a helper function for SuperfluidUndelegate and SuperfluidUndelegateToConcentratedPosition.
+// It performs the following tasks:
+// - checks that the lock is valid for superfluid staking
+// - gets the intermediary account associated with the lock id
+// - deletes the connection between the lock id and the intermediary account
+// - deletes the synthetic lockup associated with the lock id
+// - undelegates the superfluid staking position associated with the lock id and burns the underlying osmo tokens
+// - returns the intermediary account
+// if createSyntheticLockup is true, then it creates a synthetic lockup in an unlocking status for the given lock id.
+func (k Keeper) undelegateCommon(ctx sdk.Context, sender string, lockID uint64, createSyntheticLockup bool) (types.SuperfluidIntermediaryAccount, error) {
 	lock, err := k.lk.GetLockByID(ctx, lockID)
 	if err != nil {
-		return err
+		return types.SuperfluidIntermediaryAccount{}, err
 	}
 	err = k.validateLockForSF(ctx, lock, sender)
 	if err != nil {
-		return err
+		return types.SuperfluidIntermediaryAccount{}, err
 	}
 	lockedCoin := lock.Coins[0]
 
-	// get the intermediate acct asscd. with lock id, and delete the connection.
+	// get the intermediate account associated with lock id, and delete the connection.
 	intermediaryAcc, found := k.GetIntermediaryAccountFromLockId(ctx, lockID)
 	if !found {
-		return types.ErrNotSuperfluidUsedLockup
+		return types.SuperfluidIntermediaryAccount{}, types.ErrNotSuperfluidUsedLockup
 	}
 	k.DeleteLockIdIntermediaryAccountConnection(ctx, lockID)
 
-	// Delete the old synthetic lockup, and create a new synthetic lockup representing the unstaking
+	// Delete the old synthetic lockup
 	synthdenom := stakingSyntheticDenom(lockedCoin.Denom, intermediaryAcc.ValAddr)
 	err = k.lk.DeleteSyntheticLockup(ctx, lockID, synthdenom)
 	if err != nil {
-		return err
+		return types.SuperfluidIntermediaryAccount{}, err
 	}
 
 	// undelegate this lock's delegation amount, and burn the minted osmo.
 	amount, err := k.GetSuperfluidOSMOTokens(ctx, intermediaryAcc.Denom, lockedCoin.Amount)
 	if err != nil {
-		return err
+		return types.SuperfluidIntermediaryAccount{}, err
 	}
 	err = k.forceUndelegateAndBurnOsmoTokens(ctx, amount, intermediaryAcc)
 	if err != nil {
-		return err
+		return types.SuperfluidIntermediaryAccount{}, err
 	}
 
-	// Create a new synthetic lockup representing the unstaking side.
-	return k.createSyntheticLockup(ctx, lockID, intermediaryAcc, unlockingStatus)
+	if createSyntheticLockup {
```

Comment:
this logic gate seems to be newly added, can we add a test case for this?

Reply:
Logic gate was removed as per Roman's suggestion, and `SuperfluidUndelegate` creates synthetic lockup as normal while `SuperfluidUndelegateToConcentratedPosition` does not.